File: c_glib/arrow-glib/string-array.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <arrow-glib/array.hpp>
#include <arrow-glib/string-array.h>

G_BEGIN_DECLS

/**
 * SECTION: string-array
 * @short_description: UTF-8 encoded string array class
 *
 * #GArrowStringArray is a class for UTF-8 encoded string array. It
 * can store zero or more UTF-8 encoded string data.
 *
 * #GArrowStringArray is immutable. You need to use
 * #GArrowStringArrayBuilder to create a new array.
 */

G_DEFINE_TYPE(GArrowStringArray,               \
              garrow_string_array,             \
              GARROW_TYPE_BINARY_ARRAY)

static void
garrow_string_array_init(GArrowStringArray *object)
{
}

static void
garrow_string_array_class_init(GArrowStringArrayClass *klass)
{
}

/**
 * garrow_string_array_get_string:
 * @array: A #GArrowStringArray.
 * @i: The index of the target value.
 *
 * Returns: The i-th UTF-8 encoded string.
 */
gchar *
garrow_string_array_get_string(GArrowStringArray *array,
                               gint64 i)
{
  auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
  auto arrow_string_array =
    static_cast<arrow::StringArray *>(arrow_array.get());
  gint32 length;
  auto value =
    reinterpret_cast<const gchar *>(arrow_string_array->GetValue(i, &length));
  return g_strndup(value, length);
}

G_END_DECLS

File: csharp/src/Apache.Arrow/Arrays/StringArray.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;
using Apache.Arrow.Types;

namespace Apache.Arrow
{
    public class StringArray: BinaryArray, IReadOnlyList<string>, ICollection<string>
    {
        public static readonly Encoding DefaultEncoding = Encoding.UTF8;

        private Dictionary<Encoding, string[]> materializedStringStore;

        public new class Builder : BuilderBase<StringArray, Builder>
        {
            public Builder() : base(StringType.Default) { }

            protected override StringArray Build(ArrayData data)
            {
                return new StringArray(data);
            }

            public Builder Append(string value, Encoding encoding = null)
            {
                if (value == null)
                {
                    return AppendNull();
                }
                encoding = encoding ?? DefaultEncoding;
                byte[] span = encoding.GetBytes(value);
                return Append(span.AsSpan());
            }

            public Builder AppendRange(IEnumerable<string> values, Encoding encoding = null)
            {
                foreach (string value in values)
                {
                    Append(value, encoding);
                }

                return this;
            }
        }

        public StringArray(ArrayData data)
            : base(ArrowTypeId.String, data) { }

        public StringArray(int length,
            ArrowBuffer valueOffsetsBuffer,
            ArrowBuffer dataBuffer,
            ArrowBuffer nullBitmapBuffer,
            int nullCount = 0, int offset = 0)
            : this(new ArrayData(StringType.Default, length, nullCount, offset,
                new[] { nullBitmapBuffer, valueOffsetsBuffer, dataBuffer }))
        { }

        public override void Accept(IArrowArrayVisitor visitor) => Accept(this, visitor);

        /// <summary>
        /// Get the string value at the given index
        /// </summary>
        /// <param name="index">Input index</param>
        /// <param name="encoding">Optional: the string encoding, default is UTF8</param>
        /// <returns>The string object at the given index</returns>
        public string GetString(int index, Encoding encoding = default)
        {
            encoding ??= DefaultEncoding;

            if (materializedStringStore != null && materializedStringStore.TryGetValue(encoding, out string[] materializedStrings))
            {
                return materializedStrings[index];
            }

            ReadOnlySpan<byte> bytes = GetBytes(index, out bool isNull);

            if (isNull)
            {
                return null;
            }

            if (bytes.Length == 0)
            {
                return string.Empty;
            }

            unsafe
            {
                fixed (byte* data = &MemoryMarshal.GetReference(bytes))
                    return encoding.GetString(data, bytes.Length);
            }
        }

        /// <summary>
        /// Materialize the array for the given encoding to accelerate the string access
        /// </summary>
        /// <param name="encoding">Optional: the string encoding, default is UTF8</param>
        /// <remarks>This method is not thread safe when it is called in parallel with <see cref="GetString(int, Encoding)"/> or <see cref="Materialize(Encoding)"/>.</remarks>
        public void Materialize(Encoding encoding = default)
        {
            encoding ??= DefaultEncoding;

            if (IsMaterialized(encoding))
            {
                return;
            }

            if (materializedStringStore == null)
            {
                materializedStringStore = new Dictionary<Encoding, string[]>();
            }

            var stringStore = new string[Length];
            for (int i = 0; i < Length; i++)
            {
                stringStore[i] = GetString(i, encoding);
            }

            materializedStringStore[encoding] = stringStore;
        }

        /// <summary>
        /// Check if the array has been materialized for the given encoding
        /// </summary>
        /// <param name="encoding">Optional: the string encoding, default is UTF8</param>
        /// <returns>True of false whether the array has been materialized</returns>
        public bool IsMaterialized(Encoding encoding = default)
        {
            if (materializedStringStore == null)
            {
                return false;
            }

            encoding ??= DefaultEncoding;
            return materializedStringStore.ContainsKey(encoding);
        }

        int IReadOnlyCollection<string>.Count => Length;

        string IReadOnlyList<string>.this[int index] => GetString(index);

        IEnumerator<string> IEnumerable<string>.GetEnumerator()
        {
            for (int index = 0; index < Length; index++)
            {
                yield return GetString(index);
            };
        }

        IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable<string>)this).GetEnumerator();

        int ICollection<string>.Count => Length;
        bool ICollection<string>.IsReadOnly => true;
        void ICollection<string>.Add(string item) => throw new NotSupportedException("Collection is read-only.");
        bool ICollection<string>.Remove(string item) => throw new NotSupportedException("Collection is read-only.");
        void ICollection<string>.Clear() => throw new NotSupportedException("Collection is read-only.");

        bool ICollection<string>.Contains(string item)
        {
            for (int index = 0; index < Length; index++)
            {
                if (GetString(index) == item)
                    return true;
            }

            return false;
        }

        void ICollection<string>.CopyTo(string[] array, int arrayIndex)
        {
            for (int srcIndex = 0, destIndex = arrayIndex; srcIndex < Length; srcIndex++, destIndex++)
            {
                array[destIndex] = GetString(srcIndex);
            }
        }
    }
}
